torch/_functorch/_aot_autograd/subclass_utils.py

```diff
@@ -111,29 +127,83 @@ def create_subclass_meta(
 # a list of tensors that we would then need to concat together.
 # Instead, we specialize the logic for the inference vs. joint graph case.
 # NOTE: this function is hot, since we unwrap tensor subclass inputs at runtime
-def unwrap_tensor_subclasses(wrapped_args, *, is_joint_structure: bool):
-    def concat_inner_tensors_from_subclasses(xs):
-        xs_inner = []
-        for x in xs:
+
+
+def filter_symints(size: Union[Size, List[Union[int, SymInt]]]) -> List[SymInt]:
+    return [s for s in size if isinstance(s, SymInt)]
+
+
+def unwrap_tensor_subclasses(
+    wrapped_args: List[Union[Tensor, int]],
+    *,
+    append_symints: bool,
+    subclass_metas: Optional[List[Union[int, SubclassCreationMeta]]] = None,
+):
+    xs_inner: List[Union[int, Tensor, SymInt]] = []
+    sizes: List[SymInt] = []
+    for x in wrapped_args:
+        if is_traceable_wrapper_subclass(x):
+            xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
+            sizes.extend(filter_symints(x.size()))
+        else:
+            xs_inner.append(x)
+
+    if append_symints:
+        xs_inner.extend(sizes)
+    return xs_inner
+
+
+def runtime_unwrap_tensor_subclasses(
+    wrapped_args: List[Union[Tensor, int]],
+    *,
+    append_symints: bool,
+    subclass_metas: List[Union[int, SubclassCreationMeta]],
+):
+    # re-use the non-runtime version
+    xs_inner = unwrap_tensor_subclasses(
+        wrapped_args, append_symints=False, subclass_metas=None
+    )
+    if append_symints:
+        for x, meta in zip(wrapped_args, subclass_metas):
             if is_traceable_wrapper_subclass(x):
-                xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
-            else:
-                xs_inner.append(x)
-        return xs_inner
+                assert isinstance(meta, SubclassCreationMeta)
+                runtime_size = x.size()
+                symint_placeholders = meta.symint_placeholders
+                assert len(runtime_size) == len(symint_placeholders)
+                xs_inner.extend(
+                    [
+                        r
+                        for (r, is_symint) in zip(runtime_size, symint_placeholders)
+                        if is_symint
+                    ]
+                )
+    return xs_inner
+
+
+def unwrap_tensor_subclasses_maybe_joint(
```

Comment:
now that these utils are all getting pretty complicated - can we kill the `is_joint_structure` flag, and have two separate utils?


(1) a function that only handles the flat-list-of-tensors case

(2) a function that only handles the joint case (that can just call (1) twice, once for primals and once for tangents)